torch/distributed/algorithms/quantization.py

```diff
@@ -0,0 +1,130 @@
+        return input_tensor
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+def _dequantize_tensor(tensor, qtype, quant_loss=None):
+    if not isinstance(tensor, torch.Tensor):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
+            "to be a single tensor list  of type torch.Tensor."
+        )
+    if (qtype == QuantType.FP16):
+        if tensor.dtype != torch.float16:
+            raise RuntimeError(
+                "tensor dtype is {} while expected to be FP16.".format(
+                    tensor.dtype
+                )
+            )
+        elif tensor.dtype == torch.float16 and quant_loss is None:
+            return tensor.float()
+        else:
+            return tensor.float() / quant_loss
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+
+def _dequantize_tensor_list(tensors, qtype, quant_loss=None):
+    if not isinstance(tensors, list) or not all(
+        isinstance(p, torch.Tensor) for p in tensors
+    ):
+        raise RuntimeError(
+            "Invalid function argument. Expected "
+            "to be of type List[torch.Tensor]."
+        )
+    elif (qtype == QuantType.FP16):
+        tensor = [_dequantize_tensor(t, qtype) for t in tensors]
+        return tensor
+    else:
+        raise RuntimeError(
+            'Quantization type {} is not supported'.format(qtype)
+        )
+
+
+def auto_quantize(func, qtype, quant_loss=None):
+    """
+    This is a prototype API that automahically quant  the input tensors, choose the precision types, and
+    pass other necessary arguments and then dequantizes the output.
+
+    Currently it only supports:
+        . FP16 quantization method
+        . all_gather, all_to_all collective ops
+    """
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        group = kwargs.get('group', None)
+        async_op = kwargs.get('async_op', False)
+        if (async_op is True):
+            raise RuntimeError(
+                'The async_op=True mode is not supported yet.'
+            )
+        if (func == dist.all_gather):
+            tensors = args[0]
+            input_tensors = _quantize_tensor(args[1], qtype)
+            out_tensors = _quantize_tensor_list(tensors, qtype)
+            dist.all_gather(out_tensors, input_tensors, group=group, async_op=async_op)
+
+        elif (func == dist.all_to_all):
+            tensors = args[0]
+            input_tensors = _quantize_tensor_list(args[1], qtype)
+            out_tensors = _quantize_tensor_list(tensors, qtype)
+            dist.all_to_all(out_tensors, input_tensors, group=group, async_op=async_op)
+
+        else:
+            raise RuntimeError(
+                f"The collective ops {func} is not supported yet"
+            )
+
+        for i, t in enumerate(_dequantize_tensor_list(out_tensors, qtype, quant_loss=quant_loss)):
```

Comment:
why we need to enumerate and reassign the `tensors`? We can just assign the tensor to new quantized/dequantized ones in the original list, as those lists are passed as list of references anyways.

Also can we make this specific to each call? there's no `out_tensors` in other ops like `all_to_all_single`